src/compiler/compile-props.js

```diff
@@ -342,9 +329,46 @@ export function coerceProp (prop, value) {
   return coerce(value)
 }
 
-function formatType (val) {
-  return val
-    ? val.charAt(0).toUpperCase() + val.slice(1)
+function assertType (value, type) {
+  var valid
+  var expectedType
+  if (type === String) {
+    expectedType = 'string'
+    valid = typeof value === expectedType
+  } else if (type === Number) {
+    expectedType = 'number'
+    valid = typeof value === expectedType
+  } else if (type === Boolean) {
+    expectedType = 'boolean'
+    valid = typeof value === expectedType
+  } else if (type === Function) {
+    expectedType = 'function'
+    valid = typeof value === expectedType
+  } else if (type === Object) {
+    expectedType = 'object'
+    valid = isPlainObject(value)
+  } else if (type === Array) {
+    expectedType = 'array'
+    valid = isArray(value)
+  } else {
+    valid = value instanceof type
+  }
+  return {
+    valid,
+    expectedType
+  }
+}
+
+function formatTypes (types) {
+  return types.reduce(
+    (p, c, i) => { return i === 0 ? formatType(c) : `${p}, ${formatType(c)}` },
```

Comment:
Yeah the code is more readable the way you say. @yyx990803 already did that in this [commit](https://github.com/vuejs/vue/commit/daab3c5f58c97bcb7c528ae199f0419456703e8f). You are right, perhaps using ` or ` instead of `, ` is a better approach.